File: Python/Sorting Algorithm/Heap_Sort/heap_sort_using_array.py


def heapify(arr, index, n):
    i = 2 * index + 1
    flag = True
    while i <= n-1 and flag:
        if i < n - 1 and arr[i] < arr[i + 1]:
            i += 1
        if arr[i] > arr[index]:
            arr[index], arr[i] = arr[i], arr[index]
        else:
            flag = False
        index = i
        i = 2 * index + 1


def max_heap(arr, n):
    i = n // 2 - 1
    while i >= 0:
        heapify(arr, i, n)
        i -= 1


def heap_sort(arr):
    n = len(arr)
    i = n // 2 - 1
    while i >= 0:
        heapify(arr, i, n)
        i -= 1

    while n > 1:
        arr[0], arr[n - 1] = arr[n - 1], arr[0]
        n -= 1
        heapify(arr, 0, n)


if __name__ == '__main__':
    a = [12, 43, 57, 98, 34, 46, 76, 45]
    print(a)
    heap_sort(a)
    print(a)

File: C++/Sorting Algorithm/Heap Sort/heap_sort_using_array.cpp
#include<iostream>
using namespace std;

void heapify(int arr[], int index, int n){
    int i = 2 * index + 1;
    bool flag = true;
    while (i <= n-1 and flag){
        if (i < n-1 and arr[i] < arr[i + 1])
            i ++;
        if (arr[index] < arr[i])
            swap(arr[i], arr[index]);
        else
            flag = false;

        index = i;
        i = 2 * index + 1;
    }
}

void max_heap(int arr[], int n){
    for (int i = (n/2 - 1); i >= 0; i --)
        heapify(arr, i, n);
    
}

void heap_sort(int arr[], int n){
    max_heap(arr, n);
    while (n > 1){
        swap(arr[0], arr[n-1]);
        n -= 1;
        heapify(arr, 0, n);
    }
}

void print(int arr[], int n){
    for (int i = 0; i < n; i ++)
        cout << arr[i] << " ";
    cout << endl;
}

int main(){
    int *arr;
    int n;
    cout << "Enter array size :";
    cin >> n;
    arr = (int *)malloc(n * sizeof(int));
    for(int i = 0; i < n; i ++){
        cout << "Enter an Integer Element :";
        cin >> arr[i];
    }
    cout << endl << "Before Sort :" << endl;
    print(arr, n);
    cout << endl << "After sort :" << endl;
    heap_sort(arr, n);
    print(arr, n);
}